client/src/keyboard-controls.ts

```diff
@@ -0,0 +1,157 @@
+    motorSensitivity.set(motor, 1.0);
+}
+
+/** Invoked in index.tsx to add event listeners. */
+export function addKeyboardListeners(): void {
+    document.addEventListener("keydown", onKeyPress);
+    document.addEventListener("keyup", onKeyRelease);
+}
+
+export function setMotorSensitivity(motor: Motor, sensitivity: number) {
+    motorSensitivity.set(motor, sensitivity);
+}
+
+function onKeyPress(event: KeyboardEvent): void {
+    event.preventDefault();
+    if (!keyPressed(event.key)) {
+        pressedKeys[event.key] = true;
+        update();
+    }
+}
+
+function onKeyRelease(event: KeyboardEvent): void {
+    pressedKeys[event.key] = false;
+    update();
+}
+
+/**
+ * Returns true if the given key is currently pressed, false otherwise.
+ *
+ * @param key the key to examine
+ */
+function keyPressed(key: string): boolean {
+    return pressedKeys[key];
+}
+
+/**
+ * Sends commands to the rover based on user input.
+ */
+export function update(): void {
+    updateDrive();
+    updateArm();
+}
+
+/**
+ * Sends drive-related commands to the rover based on user input.
+ */
+function updateDrive(): void {
+    let leftRight: number = 0.0;
+    let forwardBackward: number = 0.0;
+    if (keyPressed("ArrowUp")) {
+        forwardBackward += 0.5;
+    }
+    if (keyPressed("ArrowDown")) {
+        forwardBackward -= 0.5;
+    }
+    if (keyPressed("ArrowLeft")) {
+        leftRight += 0.5;
+    }
+    if (keyPressed("ArrowRight")) {
+        leftRight -= 0.5;
+    }
+    if (keyPressed("c")) {
+        leftRight *= 0.5;
+        forwardBackward *= 0.5;
+    }
+    if (keyPressed("v")) {
+        leftRight *= 2.0;
+        forwardBackward *= 2.0;
+    }
+    RoverCommands.setDrivePower(forwardBackward, leftRight);
+}
+
+/**
+ * Sends arm-related commands to the rover based on user input.
+ */
+function updateArm(): void {
+    RoverCommands.setMotorPower(Motor.ARM_BASE, powerFromKeys("q", "a") * motorSensitivity.get(Motor.ARM_BASE));
+    RoverCommands.setMotorPower(Motor.SHOULDER, powerFromKeys("w", "s") * motorSensitivity.get(Motor.SHOULDER));
+    RoverCommands.setMotorPower(Motor.ELBOW, powerFromKeys("e", "d") * motorSensitivity.get(Motor.ELBOW));
+    RoverCommands.setMotorPower(Motor.FOREARM, powerFromKeys("r", "f") * motorSensitivity.get(Motor.FOREARM));
+    RoverCommands.setMotorPower(Motor.HAND, powerFromKeys("u", "j") * motorSensitivity.get(Motor.HAND));
+    updateWrist();
+}
+
+/**
+ * Returns 0.0 if neither or both keys are pressed, 1.0 if positiveKey is
+ * pressed, or -1.0 if negativeKey is pressed.
+ * 
+ * @param positiveKey the key that, when pressed, should increase the overall power
+ * @param negativeKey they that, when pressed, should decrease the overall power
+ */
+function powerFromKeys(positiveKey: string, negativeKey: string): number {
+    let power: number = 0.0;
+    if (keyPressed(positiveKey)) {
+        power += 1.0;
+    }
+    if (keyPressed(negativeKey)) {
+        power -= 1.0;
```

Comment:
An interesting effect of doing things this way is that if both keys are pressed, the power will be zero. For some reason I intuitively expected whichever key was pressed later to overwrite the earlier keypress, but this way is probably better.